### PR TITLE
Fix resolution error if inner entry doesn't provide a required API.

### PR DESCRIPTION
**Pull request by benluddy**
For bundle-to-bundle dependencies (as opposed to
subscription-to-bundle "dependencies"), candidates were filtered first
to only those that satisfy the constraint in question, then sorted in
channel order. Since the relative order of entries within a channel is
currently determined by replaces-distance from the channel head,
sorting would fail if an entry in the middle of a channel's replaces
chain did not satisfy the constraint.

For example, assume an operator "bar" has three available versions
within a single channel: bar-1, bar-2, and bar-3. If bar-1 and bar-3
provide the API "Bar", but bar-2 does not, then resolution will fail
for any operator that requires the API "Bar", because bar-1 and bar-3
are not connected by update edges without bar-2.

Now, all (package, channel, catalog) combinations satisfying a
constraint are enumerated and sorted before filtering to only those
entries that satisfy the constraint.

### CODE REVIEW

File: pkg/controller/registry/resolver/resolver_test.go [hunk 1]
```diff
@@ -1319,6 +1319,8 @@ func genOperator(name, version, replaces, pkg, channel, catalogName, catalogName
 				Namespace: catalogNamespace,
 			},
 			DefaultChannel: defaultChannel != "" && channel == defaultChannel,
+			Package:        pkg,
```
Comment (dinhxuanvu) [inline pkg/controller/registry/resolver/resolver_test.go:1322]: How are these changes relevant to the fix?

**Review by dinhxuanvu — COMMENTED** [review]
Looking good me except some changes on properties that don't seem to be relevant to the fix. Can you explain? Carry another fix within this?

**Comment by ecordell** [comment]
/approve

approach looks good and fix makes sense.

File: pkg/controller/registry/resolver/resolver.go [hunk 2]
```diff
@@ -469,6 +483,19 @@ func (r *SatResolver) newSnapshotForNamespace(namespace string, subs []*v1alpha1
 		op.sourceInfo = &OperatorSourceInfo{
 			Catalog: existingOperatorCatalog,
 		}
+		for _, p := range op.properties {
```
Comment (dinhxuanvu) [inline pkg/controller/registry/resolver/resolver.go:486]: Sorry. I should've commented here on the question how these changes relevant to the fix are.

File: test/e2e/subscription_e2e_test.go [hunk 3]
```diff
@@ -55,6 +55,74 @@ var _ = Describe("Subscription", func() {
...
+					}
+				},
+				Equal(operatorsv1alpha1.SubscriptionSpec{
+					CatalogSource:          "test-catalog",
+					CatalogSourceNamespace: testNamespace,
+					Package:                "dependency",
+					Channel:                "channel-dependency",
+				}),
+			)))
+		})
```
[the remark this answers is not in the document]
Comment (benluddy) [reply test/e2e/subscription_e2e_test.go:125]: Done.

**Review by dinhxuanvu — APPROVED** [review]
/lgtm
